core/crypto/rng_hw.c

```diff
@@ -3,19 +3,89 @@
 
 #include <compiler.h>
 #include <crypto/crypto.h>
+#include <kernel/mutex.h>
 #include <kernel/panic.h>
+#include <malloc.h>
 #include <rng_support.h>
 #include <tee/tee_cryp_utl.h>
 #include <types_ext.h>
 
-/* This is a HW RNG, no need for seeding */
+static uint8_t *rng_fifo;
+static size_t rng_fifo_size;
+static size_t rng_fifo_pos;
+static struct mutex rng_fifo_mutex = MUTEX_INITIALIZER;
+
+TEE_Result __weak hw_get_max_available_entropy(size_t *blen)
+{
+	*blen = sizeof(size_t);
+
+	return TEE_SUCCESS;
+}
+
+TEE_Result __weak hw_get_available_entropy(void *buf)
```

Comment:
I think this function should also output the number of bytes returned.
This idea of this function of not to allow driver to issue bigger burst of random byte, but rather let the function output the entropy that is available in the HW, and let caller call again to get the remaining.

So it is a bit strange to propose an implementation of `hw_get_available_entropy()` on top of `hw_get_hw_bytes()`. It should rather be the opposite.

Reply:
All the hardware I looked at produces entropy at a fixed amount of bytes per iteration. And it is either available or not. So it's easier to have `hw_get_available_entropy()` return a fixed size every time, if it is not available then return nothing. We can query for that size using `hw_get_max_available_entropy()`.

I have made an implementation in this patch that does perform `hw_get_random_bytes()` on top of `hw_get_available_entropy()`. So both ways work and are interoperable.

This `__weak` version `hw_get_available_entropy()` is here only temporary until it is supported natively by all the drivers, then we remove this temporary gasket version. This allows us in the mean time have a common non-blocking API that can implement both SMCCC TRNG and `crypto_rng_read()`.